core/discovery/broker/subscriber.go

```diff
@@ -0,0 +1,131 @@
+/*
+ * Copyright (C) 2019 The "MysteriumNetwork/node" Authors.
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License as published by
+ * the Free Software Foundation, either version 3 of the License, or
+ * (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+
+package broker
+
+import (
+	"sync"
+	"time"
+
+	"github.com/mysteriumnetwork/node/communication"
+	"github.com/mysteriumnetwork/node/communication/nats"
+	"github.com/mysteriumnetwork/node/core/discovery"
+	"github.com/mysteriumnetwork/node/market"
+)
+
+// proposalSubscriber responsible for handling proposal events through Broker (Mysterium Communication)
+type proposalSubscriber struct {
+	proposalStorage  *discovery.ProposalStorage
+	proposalIdleness time.Duration
+	receiver         communication.Receiver
+
+	watchdogStep time.Duration
+	watchdogStop chan bool
+	watchdogLock sync.Mutex
+	watchdogSeen map[market.ProposalID]time.Time
+}
+
+// NewProposalSubscriber returns new proposalSubscriber instance.
+func NewProposalSubscriber(
+	proposalsStorage *discovery.ProposalStorage,
+	proposalIdleness time.Duration,
+	connection nats.Connection,
+) *proposalSubscriber {
```

Comment:
It would be more clear if NewProposalSubscriber returns actual interface which proposalSubscriber implements.

Reply:
It doesn't in implement any interface. Rather caller defines interface while using this implementation.
Indeed made Subscriber public to improve clarity.